{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Tutorial 2 : Cleaning and Exploring BioMedical Data\n",
    "\n",
    "In this tutorial we’ll work through various features of Refine including importing data, faceting,\n",
    "clustering and calling into remote APIs by working on a fictional biomedical research project. We’ll\n",
    "start with a research question in mind and use features of Refine to gain insight and find answers.\n",
    "The research question relates to genes that are expressed in the 1st branchial arch and optic cup of\n",
    "the mouse embryo at 10.5dpc, and what diseases in humans are associated with the human\n",
    "orthologue of the mouse genes. We’ll start with a list of mouse genes expressed in the 1st branchial\n",
    "arch that we have previously extracted from a public resource (and introduced some inconsistencies\n",
    "into), correct the inconsistencies, filter the list to show only those that are also expressed in the optic\n",
    "cup, and then we’ll supplement the information in the table with both human gene synonyms, and\n",
    "human genetic disorders where there is evidence that there is an underlying molecular basis related\n",
    "to each gene.\n",
    "\n",
    "**Goal**: Find out what human genetic diseases are associated with a set of genes expressed in the mouse embryo 1st branchial arch. \n",
    "\n",
    "Start with basic set of data from the EMAGE mouse gene expression database, then use HGNC API, then OMIM API \n",
    "\n",
    "### Starting a Project\n",
    "**Step 1.** So you have some data to play with in these exercises, we’ve previously searched the EMAGE spatial gene expression database for genes expressed in the region shown in  <span style=\"color:magenta\">magenta</span>. below (i.e. a region corresponding to the 1st branchial arch of the 10.5dpc mouse embryo):<br>\n",
    "\n",
    "![tutorial2_mouse_embryo](https://user-images.githubusercontent.com/6886376/43751949-56b85250-9a42-11e8-94b1-f08a1f156352.jpg)\n",
    "\n",
    "\n",
    "We’ve exported the results in csv (i.e. tabular) format, introduced some anomalies into the\n",
    "data (so you can clean them up in this exercise), and saved them for you here: \n",
    "\n",
    "https://www.dropbox.com/s/m7c4dop0uv1qkip/Mouse%20Gene%20Data.xls?dl=1.\n",
    "\n",
    "**Step 2.** Launch Open Refine. It will open in your default web browser. Note: Open Refine is not a cloud application; it runs locally, using your web browser as its primary interface. We recommend using Firefox, Chrome or Safari. To launch OpenRefine in a browser that is not your default, you can copy the URL that appears upon launching in the default browser (e.g. 127.0.0.1:3333 in the following example), and paste this into another browser:<br>\n",
    "\n",
    "![tutorial2_ip](https://user-images.githubusercontent.com/6886376/43752119-2426ad68-9a43-11e8-883b-43408bfa30a6.jpg)<br>\n",
    "\n",
    "**Step 3.** Select the Create Project tab.<br>\n",
    "**Step 4.** Select the import data from local computer.<br>\n",
    "**Step 5.** Download the file Mouse Gene Data.xls and load this into open refine from the local computer https://www.dropbox.com/s/m7c4dop0uv1qkip/Mouse%20Gene%20Data.xls?dl=1 <br>\n",
    "\n",
    "**Step 6.** Click Next >>.\n",
    "![tutorial2_data_import](https://user-images.githubusercontent.com/6886376/43752189-5fba1b6c-9a43-11e8-8cb6-bb8a7a925c5f.jpg)<br>\n",
    "\n",
    "**Step 7.** On the resultant screen, set Project name to 10.5dpc mouse embryo expression: \n",
    "![tutorial2_data_name](https://user-images.githubusercontent.com/6886376/43752213-80d2c5d8-9a43-11e8-8ab5-82c344502ed4.jpg)\n",
    "<br>\n",
    "\n",
    "**Step 8.** Untick Parse next _1___ line(s) as column headers:\n",
    "![tutorial2_data_filter1](https://user-images.githubusercontent.com/6886376/43752262-add9a132-9a43-11e8-9d3e-bf41c6615fa8.jpg)<br>\n",
    "\n",
    "**Step 9.** Click Create Project >>.\n",
    "![tutorial2_data_filter2](https://user-images.githubusercontent.com/6886376/43752287-c8bf01a4-9a43-11e8-8c30-442324ffc39f.jpg)\n",
    "<br>\n",
    "\n",
    "**Step 10.** The project will open with 1036 rows (of which 10 are visible, and you can change to display up to 50 rows):<br>\n",
    "\n",
    "\n",
    "### Getting Organised\n",
    "Renaming columns\n",
    "**Step 1.** Select Column 1 menu > Edit column > Rename this column. \n",
    "\n",
    "![tutorial2_organised1](https://user-images.githubusercontent.com/6886376/43752331-e79af04c-9a43-11e8-975f-246481c2db12.jpg)\n",
    "<br>\n",
    "\n",
    "**Step 2.** Rename to Mouse Gene Symbol. Click OK <br>\n",
    "**Step 3.** Rename Column2 to Embryo age as above.<br>\n",
    "**Step 4.** Rename Column3 to Expression text annotation as above.<br>\n",
    "**Step 5.** Rename Column4 to EMAGE ID as above.<br>\n",
    "**Step 6.** Rename Column5 to Genotype as above.<br>\n",
    "**Step 7.** Rename Column6 to Assay type as above.<br>\n",
    "**Step 8.** Rename Column7 to Probe ID as above.<br>\n",
    "**Step 9.** Rename Column8 to Strain as above.<br>\n",
    "**Step 10.** Remove Column 9 and Column 10.<br>\n",
    "\n",
    "### Splitting columns\n",
    "**Step 1.** Select Genotype menu > Edit column > Add column based on this column...<br>\n",
    "**Step 2.** Set New column name to wt/het.<br>\n",
    "**Step 3.** Set Expression to:<br>\n",
    "if (value != \"wild-type\", \"heterozygote\", \"wt\")\n",
    "\n",
    "The GREL expression used in step 3 is being used to add values to a new column, and these values are based on the existing values found in the genotype column. The genotype column has values that are either “wild-type”, or if the embryo was a\n",
    "heterozygote, the specific genotype is shown (e.g. “Dll3<tm1Rbe>/Dll3<+>”) The first GREL expression is used to add a column that denotes if the embryos were either wild-type (wt) or heterozygote (het): <br>\n",
    "\n",
    "**GREL Expression: <span style=\"color:red\">if (value</span> <span style=\"color:blue\"> != </span> <span style=\"color:yellowgreen\"> \"wild-type\", </span> <span style=\"color:orange\">\"heterozygote\",</span> \"wt\")** <br>\n",
    "\n",
    "Meaning: for every row, <span style=\"color:red\">if the value of the cell in the genotype column</span>  <span style=\"color:blue\"> does not equal </span>  <span style=\"color:yellowgreen\">  “wild-type” </span> <span style=\"color:blue\">, then in the new column </span> <span style=\"color:orange\">\"add a value of “heterozygote”\",</span>  <span style=\"color:deeppink\">  “, otherwise in the new column” </span> **add a value of “wt”**.\n",
    "\n",
    "## Exploring the data\n",
    "### Sorting Columns\n",
    "**Step 1.** Select Genotype menu > Sort<br>\n",
    "**Step 2.** Select text and a-z and click OK. All the non wild-type will appear first.<br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "### Facet and cluster on Genes\n",
    "\n",
    "**Step 1.** Select Mouse Gene Symbol menu > Facet > Text Facet. Note that a Gene Symbol\n",
    "facet will appear on the left hand side of the screen. This shows a list of unique gene symbols\n",
    "in the data.<br>\n",
    "**Step 2.** Click on 941 choices. A text box will appear so we can copy and paste our list of unique\n",
    "gene symbols into, say, a document.<br>\n",
    "**Step 3.** Click count to order to list the most frequently occurring gene symbols first.<br>\n",
    "**Step 4.** Click Cluster to reveal and fix some gene symbol nomenclature consistency issues within the\n",
    "dataset that we’ve introduced. Select, for instance nearest neighbor as the method. You'll\n",
    "see that Refine finds some near matches. Now try some of the other methods.<br>\n",
    "**Step 5.** You can make your data more consistent by typing the correct value into New Cell Value\n",
    "and ensuring the Merge? checkbox is selected. Use the Merge Selected & Re-Cluster function to actually modify the dataset.\n",
    "\n",
    "### Undo/Redo History\n",
    "Open Refine has an infinite undo history. To access:<br>\n",
    "**Step 1.** Click on the Undo/Redo tab. You'll see every action you've done since creating the project.<br>\n",
    "**Step 2.** You can undo to any step by clicking on the step you want to revert back to. Similarly, you can redo every step. <br>\n",
    "**Step 3.** Click on Extract button. The operations you’ve performed can be saved in JSON format so that you have a record of the tasks undertaken and can apply the workflow to other similar data tables in the future.<br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filtering the list\n",
    "Because this example list is long, we can filter it and just work on a subset of the data in the list for\n",
    "the next set of operations. Please ensure you perform this step to save time during the class.<br>\n",
    "**Step 1.** Select Expression text annotation menu > Text filter<br>\n",
    "**Step 2.** Note that an Expression text annotation facet will appear on the left hand side of the screen.<br>\n",
    "**Step 3.** Type in optic and press return <br>\n",
    "**Step 4.** Note that now 17 matching rows of the original 1036 are shown that contain the word “optic”\n",
    "(and are selected for the subsequent steps).<br>\n",
    "\n",
    "### Calling into an API and parsing results\n",
    "It's relatively straightforward to draw in data from an external source with Open Refine. In this case\n",
    "we'll call in to HGNC's Human Gene Nomenclature API to get a list of all the human synonyms for the\n",
    "mouse genes in our list.<br>\n",
    "**Step 1.** Select Mouse Gene symbol menu > Edit column > Add column by fetching URLs.<br>\n",
    "**Step 2.** Type HGNC gene info into New column name.<br>\n",
    "**Step 3.** Type 20 in Throttle delay.<br>\n",
    "**Step 4.** Type the following exactly into the Expression box (including the \" symbols): <br> \"http://rest.genenames.org/fetch/symbol/\" +value\"<br>\n",
    "**Step 5.** You’ll get live feedback in the preview tab so you’ll be able to see if the syntax of the URLs looks correct:<br>\n",
    "\n",
    "\n",
    "Note :\n",
    "The HGNC API documentation is found here: http://www.genenames.org/help/rest-webservice-\n",
    "help where you can see that the **fetch** function is the main request method to retrieve particular records from the HGNC server that will return back all the stored fields. The **fetch** method requires the user to add the queriable field and the query\n",
    "term to the url (e.g. http://rest.genenames.org/fetch/symbol/ZNF3 - in this case the queriable field is ‘symbol’ and the query term is ‘ZNF3’). <br>\n",
    "\n",
    "**We can use GREL to construct these URLS for every row in our OpenRefine dataset simply by using the expression:**<br>\n",
    "\n",
    "<span style=\"color:blue\">\"http://rest.genenames.org/fetch/symbol/\"</span> <span style=\"color:red\">+</span> <span style=\"color:green\">value</span>\"<br>\n",
    "\n",
    "This is telling your computer to <span style=\"color:blue\"> visit http://rest.genenames.org/fetch/symbol/ </span> and <span style=\"color:red\">to append</span> <span style=\"color:green\">the value shown in the gene symbol column for that row</span> to the end of the url.\n",
    "\n",
    "This is the result set you should see :\n",
    "\n",
    "![tutorial2_organised2](https://user-images.githubusercontent.com/6886376/43752376-0da1de0e-9a44-11e8-8c73-3dd1857d0d41.jpg)\n",
    "\n",
    "\n",
    "**Step 6.** Click OK. Open Refine will query the HGNC API for each row in the dataset.<br>\n",
    "**Step 7.** In the new column, you should see the results in each row looking something like this:\n",
    "\n",
    "![tutorial2_organised3](https://user-images.githubusercontent.com/6886376/43752407-2e758e3c-9a44-11e8-9bbf-b99851f527b2.jpg)\n",
    "\n",
    "\n",
    "This is just html/xml but in OpenRefine due to space constraints, it isn’t formatted clearly. To see an example formatted in a structured manner, use Firefox, and paste in http://rest.genenames.org/fetch/symbol/otx1 and press return. A page like this will be seen showing the hierarchically structured file:\n",
    "\n",
    "![tutorial2_organised4](https://user-images.githubusercontent.com/6886376/43752443-5ae715f8-9a44-11e8-9d39-57a47050195f.jpg)\n",
    "\n",
    "\n",
    "We can now use GREL to parse these results to extract whatever we want from the contents of this file. We’ll extract both the official human gene symbol (approved by the Human Gene Nomenclature Committee), and the VEGA ID (VEGA is a repository for high-quality gene models produced by the manual annotation of vertebrate genomes), and add these to our table.\n",
    "\n",
    "**Step 8.** Select **HGNC gene info menu > Edit column > Add column based on this column.**<br>\n",
    "**Step 9.** Type *Human Gene Symbol* into New column name.<br>\n",
    "**Step 10.** Type parseHtml(value).select(\"str[name=symbol]\")[0].htmlText() into **Value** <br>\n",
    "\n",
    "<span style=\"color:blue\">parseHtml</span> <span style=\"color:red\">(value).select</span> (\"<span style=\"color:green\">str</span>[<span style=\"color:deeppink\">name=symbol</span>]\")<span style=\"color:orange\">[0]</span>.htmlText()\n",
    "\n",
    "This GREL command <span style=\"color:blue\">parses the html </span>and just <span style=\"color:red\">selects the value of the</span> <span style=\"color:green\">string (str)</span> with the <span style=\"color:deeppink\">name symbol</span>, <span style=\"color:orange\">returns the 1st result (if there is more than one)</span>  as html text.\n",
    "\n",
    "**Step 11.** A new column will be added showing the official human gene symbols (e.g. OTX1). Check that there is a one to one correspondence between the mouse and human gene symbols (mouse symbols are denoted in Sentence Case whereas human genes are denoted in UPPERCASE), as the Human gene symbol we have extracted will be used to search the human OMIM database at a later step.\n",
    "\n",
    "We’ll now also add the VEGA ID for the human gene<br>\n",
    "**Step 12.** Select **HGNC gene info menu > Edit column > Add column based on this column.**<br>\n",
    "**Step 13.** Type VEGA ID into **New column name.**<br>\n",
    "**Step 14.** Type parseHtml(value).select(\"str[name=vega_id]\")[0].htmlText() into **Value**<br>\n",
    "\n",
    "<span style=\"color:blue\">parseHtml</span> <span style=\"color:red\">(value).select</span> (\"<span style=\"color:green\">str</span>[<span style=\"color:deeppink\">name=vega_id</span>]\")<span style=\"color:orange\">[0]</span>.htmlText()\n",
    "\n",
    "\n",
    "This GREL command <span style=\"color:blue\">parses the html </span>and just <span style=\"color:red\">selects the value of the</span> <span style=\"color:green\">string (str)</span> with the <span style=\"color:deeppink\">name vega_id</span>, <span style=\"color:orange\">returns the 1st result (if there is more than one)</span>  as html text.\n",
    "\n",
    "**Step 15.** A new column will be added showing the VEGA IDs (e.g. OTTHUMG00000129454) if one is\n",
    "found.<br>\n",
    "**Step 16.** Select **HGNC gene info menu > Edit column > Remove this column.**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calling into another API\n",
    "This time we'll call into OMIM's API to get a list of human genetic disorders that are associated with\n",
    "our list of genes<br>\n",
    "**Step 1.** Select **Human Gene symbol menu > Edit column > Add column by fetching URLs.**<br>\n",
    "**Step 2.** Type OMIM info into **New column name.**<br>\n",
    "**Step 3.** Type 300 in **Throttle delay.**<br>\n",
    "**Step 4.** Type the following exactly into the **Expression** box:<br>\n",
    "\n",
    "\"https://api.omim.org/api/entry/search?index=clinicalSynopsis&&apiKey=oJvnufHfSQaWKo4ryl3hlw&filter=cs_molecular_basis_exists%3Atrue&search=\" + value\"\"\n",
    "\n",
    "Example :\n",
    "https://api.omim.org/api/entry/search?search=sox9&apiKey=oJvnufHfSQaWKo4ryl3hlw\n",
    " \n",
    " This is a slow API so please be patient it also throttles robot crawlers and can block your IP on occasion.\n",
    "  \n",
    " \n",
    "If you encounter an error then the OMIM API documentation is found here: http://www.omim.org/help/api. Note that there are limits in its use: “Entries and clinical synopses are limited to 20 per request if any 'includes' are specified and a API key may now be needed for this step.\n",
    "\n",
    "The rate of requests is currently limited to 4 requests per second. Your client will be throttled if you exceed the maximum rate, a 409 http status will be returned. Note that repeated requests which get a 409 will lengthen the interval before your access is restored,\n",
    "and if more than thirty 409's are generated your client will be banned for six hours after which access will be restored.”<br>\n",
    "\n",
    "In this example we want to use the OMIM API to list the genetic disorders that are associated with each gene in our list. This could be done manually on a one-by-one fashion, by visiting the OMIM advanced ‘clinical synopsis’ search page (http://www.omim.org/search/advanced/clinicalSynopsis), manually checking the ‘molecular basis’ checkbox and entering a gene symbol as a search term (e.g. Sox9) and pressing search. <br>\n",
    "\n",
    "If you try this, you’ll see the URL is structured as: <br> \n",
    "\n",
    "http://www.omim.org/search?index=clinicalSynopsis&start=1&limit=10&search=sox9&sort=score+desc&limit=10&exists=cs_molecular_basis_exists%3Atrue. <br>\n",
    "\n",
    "In this URL, everything between the ampersand (&) symbols are parameters and their order can be moved around.\n",
    "The above query can also be structured as <br>\n",
    "http://www.omim.org/search?index=clinicalSynopsis&filter=cs_molecular_basis_exists%3Atrue&search=sox9 <br>\n",
    "(and you can copy and paste this into a browser to see that the same result is returned in both cases).<br>\n",
    "\n",
    "**We can use GREL to construct URLS in this format for every row in our OpenRefine dataset simply by using the expression:**\n",
    "\n",
    "\n",
    "<span style=\"color:blue\">http://www.omim.org/</span> <span style=\"color:red\">search?</span> (\"<span style=\"color:green\">index=clinicalSynopsis</span> & <span style=\"color:deeppink\">filter=cs_molecular_basis_exists</span> %3A <span style=\"color:blue\">true</span> & <span style=\"color:orange\">search=\"+value\"</span>\n",
    "\n",
    "\"http://www.omim.org/search?index=clinicalSynopsis&filter=cs_molecular_basis_exists%3Atrue&search=\"+value\" <br>\n",
    "\n",
    "This is telling your computer to <span style=\"color:blue\"> visit http://www.omim.org/</span> <span style=\"color:red\"> and do a search</span>  <span style=\"color:green\"> of the clinical synopses data index,</span>  and to <span style=\"color:deeppink\"> filter the results to include only those where there is evidence</span>  <span style=\"color:darkblue\">(i.e.true)</span> <span style=\"color:deeppink\">  that a molecular basis exists</span> , and to <span style=\"color:orange\"> search for the gene\n",
    "symbol “value”</span> . %3A is the hex code for a colon.\n",
    "\n",
    "![](https://jamescookuniversity-my.sharepoint.com/:i:/r/personal/neil_fraser_jcu_edu_au/Documents/DBMA3405/Open%20Refine/Images/Calling%20Another%20API.PNG?csf=1&e=swU8qm)<br>\n",
    "**Step 5**  You’ll get live feedback in the preview so you’ll be able to see if the syntax of the URLs looks correct:\n",
    " \n",
    "\n",
    "**Step 6.** Click OK. Open Refine will query the API for each row in the dataset.<br>\n",
    "**Step 7.** In the new column, you should see the results in each row looking something like this:\n",
    "![tutorial2_organised6a](https://user-images.githubusercontent.com/6886376/44498298-f50cf980-a6c0-11e8-8617-1d9ee806be03.jpg)\n",
    "\n",
    "    \n",
    "We can now use GREL to parse the results to extract whatever we want from the contents of this table.<br>\n",
    "**Step 8.** Select **OMIM info > Edit column > Add column based on this column.** with column name **Disease Name**<br>\n",
    "**Step 9.** Type ***value.parseHtml().select(\"preferredTitle\")[0].htmlText()*** into Value\n",
    "**Step 10.** Select **OMIM info > Edit column > Add column based on this column.** with column name **Disease id**<br>\n",
    "**Step 11.** Type ***value.parseHtml().select(\"mimNumber\")[0].htmlText()*** into Value\n",
    "\n",
    " \n",
    "\n",
    "<span style=\"color:deeppink\">value.parseHtml().select</span> <span style=\"color:red\">(\"mimNumber\")</span> ) <span style=\"color:blue\">[0]</span> <span style=\"color:green\">.htmlText() </span> <br>\n",
    "\n",
    "\n",
    "This GREL command <span style=\"color:deeppink\">parses the html to select the value </span>a of the <span style=\"color:blue\"> 1st occurance </span> <span style=\"color:red\"> of a value ‘mimNumber’ </span> in the table, <span style=\"color:green\">as text.</span> <br>\n",
    "\n",
    "You can try changing the number in square brackets to see how this changes the result (e.g. try entering [6] instead of [0]).\n",
    "\n",
    "![](https://jamescookuniversity-my.sharepoint.com/:i:/r/personal/neil_fraser_jcu_edu_au/Documents/DBMA3405/Open%20Refine/Images/Tutorial2_organised8.jpg?csf=1)\n",
    "\n",
    "**Step 12.** two new column will be added showing Disease ID and the Disease name (e.g. # 612109.OCULOAURICULAR SYNDROME) if one is found.<br>\n",
    "\n",
    "![tutorial2_organised6](https://user-images.githubusercontent.com/6886376/44498179-65ffe180-a6c0-11e8-9837-8b871cb1cd34.jpg)\n",
    "\n",
    "**Step 13.** Select **OMIM menu > Edit column > Remove this column.**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exporting the dataset\n",
    "**Step 1.** Click **Export** in the top right-hand corner.<br>\n",
    "**Step 2.** Select **Excel** from the drop-down menu. An Excel dump of your data will be downloaded that can be opened using Excel:\n",
    "![tutorial2_export1](https://user-images.githubusercontent.com/6886376/43752659-69839a18-9a45-11e8-8c2f-08c3969dada7.jpg)<br>\n",
    "\n",
    "**Step 3.** Remember that you can also extract the operation history to keep track of how you’ve modified the data, by selecting the **Extract…** option under the **Undo/Redo** tab:\n",
    "\n",
    "![tutorial2_export2](https://user-images.githubusercontent.com/6886376/43752679-8464af2a-9a45-11e8-839c-8da18a949739.jpg)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exporting the dataset & Grel Code\n",
    "**Step 1.** Click **Export** in the top right-hand corner to export data.<br>\n",
    "**Step 2.** Select **Comma-separated variable** from the drop-down menu. A CSV dump of your data will be downloaded.<br>\n",
    "**Step 3.** To export you GREL document then open the extract tab in the left hand operations column and select all the code  for copying aross to a word or txt dovument<br>\n",
    "**Step 4.** Save both files and submit to JCU learn\n",
    "\n",
    "### Resources\n",
    "The EMAGE mouse gene expression database (source of the original dataset):\n",
    "• http://www.emouseatlas.org/emage<br>\n",
    "HGNC database of human gene names:\n",
    "• http://www.genenames.org<br>\n",
    "Online Mendelian Inheritance in Man (OMIM):\n",
    "• http://www.omim.org"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
